test/utils/norms.hpp

```diff
@@ -1,43 +1,97 @@
 #pragma once
 
-#include <finufft/test_defs.h>
-
-// ahb's low-level array helpers
-template<typename T>
-FINUFFT_EXPORT T FINUFFT_CDECL relerrtwonorm(BIGINT n, const std::complex<T> *a,
-                                             const std::complex<T> *b)
-// ||a-b||_2 / ||a||_2
-{
-  T err = 0.0, nrm = 0.0;
-  for (BIGINT m = 0; m < n; ++m) {
-    // note std::norm here & below is |a|^2 ("field norm") not usual |a| ...
-    nrm += std::norm(a[m]);
-    err += std::norm(a[m] - b[m]);
+#include <algorithm>
+#include <cmath>
+#include <complex>
+#include <type_traits>
+#ifdef __CUDACC__
+#include <thrust/complex.h>
+#endif
+
+// ----------------------------------------------------------------------------
+// Compute ||a–b||₂ / ||a||₂
+// ----------------------------------------------------------------------------
+template<typename BIGINT,
+         typename ArrA, // supports ArrA[i] -> complex-like
+         typename ArrB  // supports ArrB[i] -> same complex-like
+         >
+auto relerrtwonorm(BIGINT n, ArrA aArr, ArrB bArr) -> decltype(aArr[0].real()) {
+  using Complex = std::decay_t<decltype(aArr[0])>;
+  using FLT     = decltype(aArr[0].real());
+
+  static_assert(std::is_floating_point<FLT>::value,
+                "relerrtwonorm: value_type must be floating point");
+  static_assert(std::is_same<Complex, std::decay_t<decltype(bArr[0])>>::value,
+                "relerrtwonorm: array element types must match");
+
+  FLT err = FLT{0}, nrm = FLT{0};
+  for (BIGINT i = 0; i < n; ++i) {
+    auto ai = aArr[i], bi = bArr[i];
+    FLT dr = ai.real() - bi.real(), di = ai.imag() - bi.imag();
+    err += dr * dr + di * di;
```

Comment:
Why is this needed, vs the old code with std::norm?  If it's for CUDA, please add comments to explain... thanks.